app/controllers/comments_controller.rb

```diff
@@ -3,73 +3,73 @@
 #   the COPYRIGHT file.
 
 class CommentsController < ApplicationController
-  include ApplicationHelper
-  before_action :authenticate_user!, :except => [:index]
+  before_action :authenticate_user!, except: :index
 
-  respond_to :html,
-             :mobile,
-             :json
+  respond_to :html, :mobile, :json
 
   rescue_from ActiveRecord::RecordNotFound do
-    render :nothing => true, :status => 404
+    render nothing: true, status: 404
   end
 
   def create
-    post = current_user.find_visible_shareable_by_id(Post, params[:post_id])
-    @comment = current_user.comment!(post, params[:text]) if post
-
+    @comment = CommentCreationService
+                 .new(post_id: params[:post_id], text: params[:text], user: current_user).create_comment!
     if @comment
-      respond_to do |format|
-        format.json{ render :json => CommentPresenter.new(@comment), :status => 201 }
-        format.html{ render :nothing => true, :status => 201 }
-        format.mobile{ render :partial => 'comment', :locals => {:post => @comment.post, :comment => @comment} }
-      end
+      respond_create_success
     else
-      render :nothing => true, :status => 422
+      render nothing: true, status: 422
     end
   end
 
   def destroy
-    @comment = Comment.find(params[:id])
-    if current_user.owns?(@comment) || current_user.owns?(@comment.parent)
-      current_user.retract(@comment)
-      respond_to do |format|
-        format.js { render :nothing => true, :status => 204 }
-        format.json { render :nothing => true, :status => 204 }
-        format.mobile{ redirect_to :back }
-      end
+    deletion_service = CommentDeletionService.new(comment_id: params[:id], user: current_user)
+    if deletion_service.allow_destroy?
+      deletion_service.destroy_comment
```

Comment:
Since we don't differentiate between not allowed and other failures, I'd say move the allowance check into `destroy_comment` and return `true`/`false` from it.